### PR TITLE
feat: implemented retry function

### PR DESCRIPTION
**Pull request by shaffeeullah**

### CODE REVIEW

File: src/storage.ts [hunk 1]
```diff
@@ -413,10 +414,42 @@ export class Storage extends Service {
...
+          return true;
+        }
+
+        if (err.errors) {
+          for (const e of err.errors) {
+            const reason = e.reason;
+            if (
+              reason === 'rateLimitExceeded' ||
+              reason === 'userRateLimitExceeded' ||
+              (reason && reason.includes('EAI_AGAIN')) ||
```
Comment (shaffeeullah) [inline src/storage.ts:436]: @frankyn The 3 above errors were already being retried via common, but are not listed in the retry guidance document. Should they continue to be retried? Or should I remove these?

File: src/storage.ts [hunk 2]
```diff
@@ -413,10 +414,42 @@ export class Storage extends Service {
...
+              reason === 'Connection Reset By Peer' ||
+              reason === 'Unexpected Connection Closure'
```
Comment (shaffeeullah) [inline src/storage.ts:438]: @frankyn What would be a good way to force these failures to see if this is the correct string to be looking for?

Comment (frankyn) [reply src/storage.ts:442]: Emulator has Reset connection and Unexpected connection closure support. We can talk through this in our meeting later today.

File: src/storage.ts (hunk 1 above)
Comment (frankyn) [reply src/storage.ts:440]: Chime in your doc

File: src/storage.ts [hunk 3]
```diff
@@ -413,10 +418,42 @@ export class Storage extends Service {
...
     const baseUrl = EMULATOR_HOST || `${options.apiEndpoint}/storage/v1`;
 
+    /**
+     * Returns true if the API request should be retried, given the error that was
+     * given the first time the request was attempted.
+     *
+     * @param {error} err - The API error to check if it is appropriate to retry.
+     * @return {boolean} True if the API request should be retried, false otherwise.
+     */
+    const retryFunction = function (err?: ApiError) {
```
Comment (shaffeeullah) [inline src/storage.ts:428]: TODO: make this configurable

**Review by JustinBeckwith — COMMENTED** [review]
would like us to be super careful here.  A few thoughts!

There is already logic to retry requests built into nodejs-common, where most of these calls go through.  It's here:  https://github.com/googleapis/nodejs-common/blob/master/src/util.ts#L513

Can we use a single implementation?

I also want to dig in on consistency across clients.  If this is implemented one way in Storage, and another in PubSub, we're going to run into trouble.  Is there a design doc folks who do nodejs can jointly work on and chat about this?

**Comment by JustinBeckwith** [comment]
@shaffeeullah and I took this offline, and I am onboard and happy 👍

File: src/storage.ts [hunk 4]
```diff
@@ -413,10 +418,42 @@ export class Storage extends Service {
...
+      if (err) {
+        if ([408, 429, 500, 502, 503, 504].indexOf(err.code!) !== -1) {
+          return true;
+        }
+
+        if (err.errors) {
+          for (const e of err.errors) {
+            const reason = e.reason;
+            if (
+              reason === 'rateLimitExceeded' ||
```
[the remark this answers is not in the document]
Comment (shaffeeullah) [reply src/storage.ts:438]: This will be removed. Spoke with frank about this offline, this error won't be thrown by GCS

File: src/storage.ts [hunk 5]
```diff
@@ -218,6 +218,36 @@ const TOTAL_TIMEOUT_DEFAULT = 600;
...
+ * @param {error} err - The API error to check if it is appropriate to retry.
+ * @return {boolean} True if the API request should be retried, false otherwise.
+ */
+const RETRYABLE_ERR_FN_DEFAULT = function (err?: ApiError) {
+  if (err) {
+    if ([408, 429, 500, 502, 503, 504].indexOf(err.code!) !== -1) {
+      return true;
+    }
+
+    if (err.errors) {
```
Comment (ddelgrosso1) [inline src/storage.ts:235]: Is there a situation where ```err !== undefined``` but ```err.errors === undefined``` (or really anything non-enumerable)?

Comment (shaffeeullah) [reply src/storage.ts:235]: I dont believe so. If there were, the bug would have already surfaced [here](https://github.com/googleapis/nodejs-common/blob/56764dc58634d151f27b67869d38a7ef97708b6c/src/util.ts#L513)

Comment (ddelgrosso1) [reply src/storage.ts:235]: I think it would be safe to remove this ```if``` statement in that case.

Comment (shaffeeullah) [reply src/storage.ts:235]: Sorry I misread your original question. It is possible for err.errors to be undefined. For example, in this test it is undefined. https://github.com/googleapis/nodejs-storage/blob/cec5e15780a0fb726247f4c281391157bdaa1cda/test/index.ts#L307

Comment (ddelgrosso1) [reply src/storage.ts:235]: I see it now: https://github.com/googleapis/nodejs-common/blob/4a1b1ab2bdedb92e4c5b102f04b005c252887c63/src/util.ts#L228. Thanks for clarifying!

**Review by ddelgrosso1 — APPROVED** [review]
LGTM

**Pull request merged**